File: src/NodeParser/NewExpressionParser.ts
import ts from "typescript";
import type { NodeParser } from "../NodeParser.js";
import { Context } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { UnknownNodeError } from "../Error/Errors.js";

export class NewExpressionParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
    ) {}

    public supportsNode(node: ts.NewExpression): boolean {
        return node.kind === ts.SyntaxKind.NewExpression;
    }

    public createType(node: ts.NewExpression, context: Context): BaseType {
        const type = this.typeChecker.getTypeAtLocation(node);

        const symbol = type.symbol || type.aliasSymbol;

        const decl =
            this.typeChecker.typeToTypeNode(type, node, ts.NodeBuilderFlags.IgnoreErrors) ||
            symbol?.valueDeclaration ||
            symbol?.declarations?.[0];

        if (!decl) {
            throw new UnknownNodeError(node);
        }

        return this.childNodeParser.createType(decl, this.createSubContext(node, context));
    }

    protected createSubContext(node: ts.NewExpression, parentContext: Context): Context {
        const subContext = new Context(node);

        if (node.arguments) {
            for (const arg of node.arguments) {
                const type = this.childNodeParser.createType(arg, parentContext);
                subContext.pushArgument(type);
            }
        }
        return subContext;
    }
}

File: internal/parser/new_expression_parser.go
package parser

// Port of src/NodeParser/NewExpressionParser.ts.

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

type NewExpressionParser struct {
	typeChecker     *checker.Checker
	childNodeParser NodeParser
}

func NewNewExpressionParser(typeChecker *checker.Checker, childNodeParser NodeParser) *NewExpressionParser {
	return &NewExpressionParser{typeChecker: typeChecker, childNodeParser: childNodeParser}
}

func (p *NewExpressionParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindNewExpression
}

func (p *NewExpressionParser) CreateType(node *ast.Node, context *Context, _ *types.ReferenceType) types.Type {
	t := p.typeChecker.GetTypeAtLocation(node)

	symbol := t.Symbol()
	if symbol == nil && t.Alias() != nil {
		symbol = t.Alias().Symbol()
	}

	decl := p.typeChecker.TypeToTypeNode(t, node, nodeBuilderFlagsIgnoreErrors, synthesizedSymbols)
	if decl == nil && symbol != nil {
		decl = symbol.ValueDeclaration
		if decl == nil && len(symbol.Declarations) > 0 {
			decl = symbol.Declarations[0]
		}
	}

	if decl == nil {
		panic(NewUnknownNodeError(node))
	}

	return p.childNodeParser.CreateType(decl, p.createSubContext(node, context), nil)
}

func (p *NewExpressionParser) createSubContext(node *ast.Node, parentContext *Context) *Context {
	subContext := NewContext(node)
	for _, arg := range node.Arguments() {
		subContext.PushArgument(p.childNodeParser.CreateType(arg, parentContext, nil))
	}
	return subContext
}
